gtsam/sfm/SfmData.cpp

```diff
@@ -0,0 +1,408 @@
+/* ----------------------------------------------------------------------------
+
+ * GTSAM Copyright 2010, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * Authors: Frank Dellaert, et al. (see THANKS for the full author list)
+
+ * See LICENSE for the license information
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file SfmData.cpp
+ * @date January 2022
+ * @author Frank dellaert
+ * @brief Data structure for dealing with Structure from Motion data
+ */
+
+#include <gtsam/inference/Symbol.h>
+#include <gtsam/sfm/SfmData.h>
+
+#include <fstream>
+#include <iostream>
+
+namespace gtsam {
+
+using std::cout;
+using std::endl;
+
+using gtsam::symbol_shorthand::P;
+using gtsam::symbol_shorthand::X;
+
+/* ************************************************************************** */
+void SfmData::print(const std::string &s) const {
+  std::cout << "Number of cameras = " << nrCameras() << std::endl;
+  std::cout << "Number of tracks = " << nrTracks() << std::endl;
+}
+
+/* ************************************************************************** */
+bool SfmData::equals(const SfmData &sfmData, double tol) const {
+  // check number of cameras and tracks
+  if (nrCameras() != sfmData.nrCameras() || nrTracks() != sfmData.nrTracks()) {
+    return false;
+  }
+
+  // check each camera
+  for (size_t i = 0; i < nrCameras(); ++i) {
+    if (!camera(i).equals(sfmData.camera(i), tol)) {
+      return false;
+    }
+  }
+
+  // check each track
+  for (size_t j = 0; j < nrTracks(); ++j) {
+    if (!track(j).equals(sfmData.track(j), tol)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+/* ************************************************************************* */
+Rot3 openGLFixedRotation() {  // this is due to different convention for
+                              // cameras in gtsam and openGL
+  /* R = [ 1   0   0
+   *       0  -1   0
+   *       0   0  -1]
+   */
+  Matrix3 R_mat = Matrix3::Zero(3, 3);
+  R_mat(0, 0) = 1.0;
+  R_mat(1, 1) = -1.0;
+  R_mat(2, 2) = -1.0;
+  return Rot3(R_mat);
+}
+
+/* ************************************************************************* */
+Pose3 openGL2gtsam(const Rot3 &R, double tx, double ty, double tz) {
+  Rot3 R90 = openGLFixedRotation();
+  Rot3 wRc = (R.inverse()).compose(R90);
+
+  // Our camera-to-world translation wTc = -R'*t
+  return Pose3(wRc, R.unrotate(Point3(-tx, -ty, -tz)));
+}
+
+/* ************************************************************************* */
+Pose3 gtsam2openGL(const Rot3 &R, double tx, double ty, double tz) {
```

Comment:
Random comment on my end: I don't love the `x2y()` functions, since it makes me think x -> y, but from our Pose3 notation it could also mean x <- y. But out of the scope of this PR, so feel free to ignore.

Reply:
I agree, but indeed out of scope